ocaml/xapi/xapi_pool.ml

```diff
@@ -416,6 +416,22 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
       raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["Primary address type differs"]));
   in
 
+  let assert_compatible_network_purpose () = try (
+    let my_nbdish =
+      Db.Network.get_all ~__context |>
+      List.map (fun nwk -> Db.Network.get_purpose ~__context ~self:nwk) |>
+      List.flatten |>
+      List.find (function `nbd | `insecure_nbd -> true | _ -> false) in
+    let remote_nbdish =
+      Client.Network.get_all rpc session_id |>
+      List.map (fun nwk -> Client.Network.get_purpose ~rpc ~session_id ~self:nwk) |>
+      List.flatten |>
+      List.find (function `nbd | `insecure_nbd -> true | _ -> false) in
+    if remote_nbdish <> my_nbdish then
+      raise Api_errors.(Server_error(operation_not_allowed, ["Incompatible network purposes: nbd and insecure_nbd"]))
```

Comment:
If I'm not mistaken, this function does not check for conflict between `nbd` and `insecure_nbd`, because in the above functions we don't treat them differently. We could use ``List.exists ((=) `nbd)`` and the same for `insecure_nbd` to detect conflicts. Maybe it would be worth making an `assert_network_purposes_compatible purposes` function, which decides whether the set of purposes is compatible or not, and we could call that from here with the union of the two networks' purposes, and from `assert_can_add_purpose` with the current set of purposes + the new purpose? Currently this is all the checking we need, inspecting the set of all purposes regardless of the networks themselves. Maybe in the future we may need to add more specialised code to one or both of these functions, but for now checking the set of purposes should be fine & simple in my opinion.

Reply:
It would be useful to have a comment somewhere that explains what the rules are for these things. At most one network in the pool can have `(nbd || insecure_nbd)`?

Reply:
Apart from what @gaborigloi said, I think that the rules is also sometimes too strict. Shouldn't we allow adding a host without an nbd network, even if the pool has one? I think that we should check that _if_ a network has an explicit purpose on the joining host, that it is the same as on the pool.

Reply:
The only rules is that it's not allowed to have both `nbd` and `insecure_nbd` networks. It's allowed to have hosts without `nbd` networks.

Reply:
It is allowed to have a pool with multiple `nbd` networks, or multiple `insecure_nbd` networks (if you don't have both types)?

Reply:
Yes, it is allowed - I think if we didn't allow it that would be too inflexible.

Reply:
@gaborigloi wrote,
> If I'm not mistaken, this function does not check for conflict between `nbd` and `insecure_nbd`,

I believe you're mistaken the check is in `if remote_nbdish <> my_nbdish then`.
Evidently the code isn't clear enough in what it's doing though.
I'll add a comment:
"A pool is not allowed both an `nbd and an `insecure_nbd network.
Therefore look for an `nbd or `insecure_nbd network locally, then remotely, then fail if they differ. If in either pool we don't find one, all is well."

Reply:
@robhoes wrote:
> Shouldn't we allow adding a host without an nbd network, even if the pool has one? I think that we should check that if a network has an explicit purpose on the joining host, that it is the same as on the pool.

That's exactly what this code does, surely?

Reply:
Sorry, I misread the code, it works fine. We should make sure that in the future we won't have a large set of duplicated checks here & in `assert_can_add_purpose`, but we discussed that it's fine to deal with that in the future when we spot some patterns that we can factor out.

Reply:
@thomassa Right you are!